### PR TITLE
ML-2286: Allow PyArrow to set parquet schema

### PR DESCRIPTION
**Pull request by gtopper**
when columns and infer_columns_from_data are not passed.

### CODE REVIEW

File: storey/targets.py [hunk 1]
```diff
@@ -299,6 +304,9 @@ class CSVTarget(_Batching, _Writer):
 
     def __init__(self, path: str, columns: Optional[List[str]] = None, header: bool = False, infer_columns_from_data: Optional[bool] = None,
                  max_lines_before_flush: int = 128, max_seconds_before_flush: int = 3, **kwargs):
+        if not columns:
```
Comment (assaf758) [inline storey/targets.py:306]: what if columns is not provided and user's infer_columns_from_data is false - do we silently overwrite infer_columns_from_data to true?

File: storey/targets.py [hunk 2]
```diff
@@ -378,8 +386,9 @@ class ParquetTarget(_Batching, _Writer):
         Optional. Defaults to None (will be inferred if event is dictionary).
     :param infer_columns_from_data: Whether to infer columns from the first event, when events are dictionaries.
         If True, columns will be inferred from data and used in place of explicit columns list if none was provided, or appended to the
-        provided list. If header is True and columns is not provided, infer_columns_from_data=True is implied.
-        Optional. Default to False if columns is provided, True otherwise.
+        provided list. If columns were not provided and infer_columns_from_data=False, PyArrow will infer the schema per file written,
+        which may cause the schemas to differ between files (e.g. if a column in all null in one file but not in another).
```
Comment (assaf758) [inline storey/targets.py]: in all => is all

File: storey/targets.py [hunk 3]
```diff
@@ -214,6 +211,12 @@ def _get_column_data_from_dict(self, new_data, event, columns, columns_types, me
                     new_data.append(new_value)
                 else:
                     new_data[column] = new_value
+            return new_data
```
Comment (assaf758) [inline storey/targets.py]: why do you need a return new_data? btw, no caller of this function use the returned object..

File: storey/targets.py [hunk 4]
```diff
@@ -450,9 +460,12 @@ def _event_to_batch_entry(self, event):
 
     async def _emit(self, batch, batch_key, batch_time, last_event_time=None):
         df_columns = []
-        if self._index_cols:
-            df_columns.extend(self._index_cols)
-        df_columns.extend(self._non_partition_columns)
+        if self._non_partition_columns:
```
Comment (assaf758) [inline storey/targets.py:462]: is this related to the infer_columns_from_data change or additional unrelated change?

File: storey/targets.py (hunk 1 above)
Comment (gtopper) [reply storey/targets.py:306]: We do, as is documented under `:param infer_columns_from_data:`.

File: storey/targets.py (hunk 4 above)
Comment (gtopper) [reply storey/targets.py:462]: It's related. Without this change, you may get a schema with only the index columns, which we do not want.

File: storey/targets.py (hunk 3 above)
Comment (gtopper) [reply storey/targets.py]: No need. Refactoring accident.

**Review by assaf758 — APPROVED** [review]
